IDEMPIERE-6356 Theme Selector on Login

CarlosRuiz-globalqss opened the pull request:
https://idempiere.atlassian.net/browse/IDEMPIERE-6356

# Pull Request Checklist

- [x] My code follows the [code guidelines](https://wiki.idempiere.org/en/Contributing_to_Trunk) of this project
- [x] My code follows the best practices of this project
- [x] I have performed a self-review of my own code
- [x] My code is easy to understand and review. 
- [x] I have checked my code and corrected any misspellings
- [x] In hard-to-understand areas, I have commented my code.
- [x] My changes generate no new warnings
### Tests
- [x] I have tested the direct scenario that my code is solving
- [x] I checked all collaterals that can be affected by my changes, and tested other potential affected scenarios
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added unit tests that prove my fix is effective or that my feature works
### Documentation
- [ ] I have made corresponding changes to the documentation as follows:
- - [ ] Improvement (improves and existing functionality): This documentation is needed if the improvement changes the way the user interacts with the system or the outcome of a process/task changes. If it is just, for instance, a performance improvement, documentation might not be needed. 
- [ ] The changed/added documentation is in the project wiki (not privately-hosted pdf files or links pointing to a company website) and is complete and self-explanatory.

CarlosRuiz-globalqss commented:
Hi @hengsin - I have a doubt with this pull request.

Was there a reason why org.idempiere.zk.breeze.theme/src/web/theme/default/preference.zul was setting the zktheme cookie name as "breeze" instead of default?

I'm setting the pull request in draft because I think if approved, we better remove the migration script and keep just this as an option for implementors to configure, not as default, mostly because the "breeze" theme is deprecated and I think we must not encourage its usage.

hengsin commented:
> Hi @hengsin - I have a doubt with this pull request.
> 
> Was there a reason why org.idempiere.zk.breeze.theme/src/web/theme/default/preference.zul was setting the zktheme cookie name as "breeze" instead of default?
> 
> I'm setting the pull request in draft because I think if approved, we better remove the migration script and keep just this as an option for implementors to configure, not as default, mostly because the "breeze" theme is deprecated and I think we must not encourage its usage.

Don't think it will work if you use "default" instead. As it is, iDempiere theme is implemented as customization of zk theme and the underlying zk theme for "default" is "breeze".

CarlosRuiz-globalqss commented:
> Don't think it will work if you use "default" instead. As it is, iDempiere theme is implemented as customization of zk theme and the underlying zk theme for "default" is "breeze".

Thanks @hengsin - in test-feature it's working fine, what could be the bad consequences?

I tried also renaming default to breeze, but that seems like a very big change and I think you didn't want to do it because it can broke other things.

I can hardcode that when reading the cookie "breeze" it will set the theme to default, but not sure if that's a good idea instead of what's already working.

CarlosRuiz-globalqss commented:
Hmm no - it seems is not working 100% - I'll do another check

CarlosRuiz-globalqss commented:
OK, in last commit I left the preference.zul back to save the zktheme cookie as breeze and controlled the situation at org.adempiere.webui.panel.LoginPanel.initComponents

CarlosRuiz-globalqss commented:
Hi @hengsin - please check my last commit:

I changed org.adempiere.ui.zk/WEB-INF/zk.xml - I think maybe it was something forgotten from the change to iceblue_c

I also added a comment in org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/theme/ThemeManager
do you think is worthy to change the code to use Themes.hasTheme instead?

Also, I'll keep this PR in draft until the finding from Diego is solved, I'll add a comment in the ticket about that.

hengsin commented:
> Hi @hengsin - please check my last commit:
> 
> I changed org.adempiere.ui.zk/WEB-INF/zk.xml - I think maybe it was something forgotten from the change to iceblue_c
> 
> I also added a comment in org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/theme/ThemeManager do you think is worthy to change the code to use Themes.hasTheme instead?
> 
> Also, I'll keep this PR in draft until the finding from Diego is solved, I'll add a comment in the ticket about that.

Yes, forgot to change that. Perhaps that should be removed now you want to use cookies only.

CarlosRuiz-globalqss commented:
> Yes, forgot to change that. Perhaps that should be removed now you want to use cookies only.

Ah OK, removed.

I tried that and still doesn't solve the issue described in the ticket.

hengsin commented on org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/theme/ThemeManager.java:
```diff
@@ -79,11 +79,14 @@ public static String getTheme() {
+					// NOTE: Here we could use also Themes.hasTheme instead, but be careful that default is registered as breeze
 					// Verify the theme.css.dsp exists in the theme folder
 					String themeCSSURL = getStyleSheet(theme);
 					if (ThemeManager.class.getResource(toClassPathResourcePath(themeCSSURL)) == null) {
```
to implement cookie base user theme selection, we need to change ThemeManager.getTheme to get from cookie (if available) as first priority. We can't use Theme.setTheme to change iDempiere theme, that wouldn't work.
